crates/core/src/factories/core_fetch.rs

```diff
@@ -0,0 +1,328 @@
+        Box::pin(async move {
+            // Add ops to set.
+            let ops = &mut self.0.state.lock().await.ops;
+            ops.extend(
+                op_list
+                    .clone()
+                    .into_iter()
+                    .map(|op_id| (op_id.clone(), source.clone())),
+            );
+
+            // Pass ops to fetch tasks.
+            op_list.into_iter().for_each(|op_id| {
+                if let Err(err) =
+                    self.0.fetch_queue_tx.try_send((op_id, source.clone()))
+                {
+                    tracing::warn!(
+                        "could not pass fetch request to fetch task: {err}"
+                    );
+                }
+            });
+
+            Ok(())
+        })
+    }
+}
+
+type FetchRequest = (OpId, AgentId);
+
+#[derive(Debug)]
+struct State {
+    ops: HashSet<FetchRequest>,
+    cool_down_list: CoolDownList,
+}
+
+#[derive(Debug)]
+struct Inner {
+    state: Arc<Mutex<State>>,
+    fetch_queue_tx: Sender<FetchRequest>,
+    fetch_queue_rx: Arc<Mutex<Receiver<FetchRequest>>>,
+}
+
+impl Inner {
+    pub fn new(cool_down_interval_ms: u64) -> Self {
+        // Create a channel to send new ops to fetch to the tasks. This is in effect the fetch queue.
+        let (fetch_queue_tx, fetch_queue_rx) = channel::<FetchRequest>(1024);
+        let fetch_queue_rx = Arc::new(Mutex::new(fetch_queue_rx));
+
+        let ops = HashSet::new();
+        let cool_down_list = CoolDownList::new(cool_down_interval_ms);
+
+        Self {
+            state: Arc::new(Mutex::new(State {
+                ops,
+                cool_down_list,
+            })),
+            fetch_queue_tx,
+            fetch_queue_rx,
+        }
+    }
+
+    pub fn spawn_fetch_tasks(
+        &self,
+        parallel_request_count: u8,
+        transport: DynTransport,
+    ) {
+        for _ in 0..parallel_request_count {
+            tokio::task::spawn(Inner::fetch_task(
+                self.fetch_queue_tx.clone(),
+                self.fetch_queue_rx.clone(),
+                transport.clone(),
+                self.state.clone(),
+            ));
+        }
+    }
+
+    async fn fetch_task(
+        fetch_request_tx: Sender<FetchRequest>,
+        fetch_request_rx: Arc<Mutex<Receiver<FetchRequest>>>,
+        transport: DynTransport,
+        state: Arc<Mutex<State>>,
+    ) {
+        while let Some((op_id, agent_id)) =
+            fetch_request_rx.lock().await.recv().await
+        {
+            // Ensure op is still in the set of ops to fetch.
+            if !state
+                .lock()
+                .await
+                .ops
+                .contains(&(op_id.clone(), agent_id.clone()))
+            {
+                continue;
+            }
+
+            // Check if agent is on cool-down list.
+            if !state
+                .lock()
+                .await
+                .cool_down_list
+                .is_agent_cooling_down(&agent_id)
```

Comment:
Only lock once (assuming the switch to std::sync::Mutex is made):

```suggestion
            let is_agent_cooling_down = {
                // It is okay to unwrap std::sync::Mutexes
                // we *want* poisoned threads to panic
                // their calling threads!
                let lock = state.lock().unwrap();
                if lock.ops.contains(&(op_id.clone(), agent_id.clone())) {
                    continue;
                }
                lock.cool_down_list.is_agent_cooling_down(&agent_id)
            };
```

Reply:
also done https://github.com/holochain/kitsune2/pull/35/commits/9add269447278ecaf2a1e4a4fad8142cf8c61011